canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,899 @@
+
+void pushKeyFrameMetrics(Frame frame, CustomData *cusData)
+{
+    updateFragmentEndTimes(frame.presentationTs, cusData->lastKeyFrameTime, cusData->timeOfNextKeyFrame);
+    
+    Aws::CloudWatch::Model::MetricDatum metricDatum;
+    Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+    cwRequest.SetNamespace("KinesisVideoSDKCanaryCPP");    
+
+    auto stream_metrics = cusData->kinesis_video_stream->getMetrics();
+    auto client_metrics = cusData->kinesis_video_stream->getProducer().getMetrics();
+    auto stream_metrics_raw = stream_metrics.getRawMetrics();
+    
+    double frameRate = stream_metrics.getCurrentElementaryFrameRate();
+    pushMetric("FrameRate", frameRate, Aws::CloudWatch::Model::StandardUnit::Count_Second, metricDatum, cusData->pDimension_per_stream, cwRequest);
+    LOG_DEBUG("Frame Rate: " << frameRate);
+
+    double transferRate = 8 * stream_metrics.getCurrentTransferRate() / 1024; // *8 makes it bytes->bits. /1024 bits->kilobits
+    pushMetric("TransferRate", transferRate, Aws::CloudWatch::Model::StandardUnit::Kilobits_Second, metricDatum, cusData->pDimension_per_stream, cwRequest);
+    LOG_DEBUG("Transfer Rate: " << transferRate);
+
+    double currentViewDuration = stream_metrics.getCurrentViewDuration().count();
+    pushMetric("CurrentViewDuration", currentViewDuration, Aws::CloudWatch::Model::StandardUnit::Milliseconds, metricDatum, cusData->pDimension_per_stream, cwRequest);
+    LOG_DEBUG("Current View Duration: " << currentViewDuration);
+
+    double availableStoreSize = client_metrics.getContentStoreSizeSize() / 1000; // [kilobytes]
+    pushMetric("ContentStoreAvailableSize", availableStoreSize, Aws::CloudWatch::Model::StandardUnit::Kilobytes, metricDatum, cusData->pDimension_per_stream, cwRequest);
+    LOG_DEBUG("Content Store Available Size: " << availableStoreSize);
+
+    if (cusData->pCanaryConfig->useAggMetrics)
+    {
+        pushMetric("FrameRate", frameRate, Aws::CloudWatch::Model::StandardUnit::Count_Second, metricDatum, cusData->pAggregated_dimension, cwRequest);
+        pushMetric("TransferRate", transferRate, Aws::CloudWatch::Model::StandardUnit::Kilobits_Second, metricDatum, cusData->pAggregated_dimension, cwRequest);
+        pushMetric("CurrentViewDuration", currentViewDuration, Aws::CloudWatch::Model::StandardUnit::Milliseconds, metricDatum, cusData->pAggregated_dimension, cwRequest);
+        pushMetric("ContentStoreAvailableSize", availableStoreSize, Aws::CloudWatch::Model::StandardUnit::Kilobytes, metricDatum, cusData->pAggregated_dimension, cwRequest);
+    }
+
+    // Capture error rate metrics every 60 seconds
+    double duration = duration_cast<seconds>(system_clock::now().time_since_epoch()).count() - cusData->timeCounter;
+    if(duration > 60)
+    {
+        cusData->timeCounter = duration_cast<seconds>(system_clock::now().time_since_epoch()).count();
+
+        double newPutFrameErrors = (double)stream_metrics_raw->putFrameErrors - cusData->totalPutFrameErrorCount;
```

Comment:
Number of errors need not be double. It can be UINT64. Only rate needs to be DOUBLE.